### PR TITLE
Finatra

### PR DESCRIPTION
**Pull request by franklinhu**
Port `zipkin-web` from Rails to Scala (using Finatra) for a variety of reasons
- Simpler deployment (all JVM-based)
- Simpler web logic (no need to replicate all the common data classes)
- Operational transparency

Changes:
- JsonAdapter and JsonQueryAdapter
  - necessary since Javascript doesn't really support Longs, so we need to convert them to strings

### CODE REVIEW

**Comment by franklinhu** [comment]
The vast majority of the js/css is copied over from zipkin-web and doesn't need to be reviewed. It's only there because sbt doesn't seem to package symlinked resources.

File: project/Project.scala [hunk 1]
```diff
@@ -275,6 +280,31 @@ object Zipkin extends Build {
...
+      version := "0.2.0-SNAPSHOT",
+      resolvers += "finatra" at "http://repo.juliocapote.com",
+      resolvers += "codahale" at "http://repo.codahale.com",
+
+      libraryDependencies ++= Seq(
+        "com.posterous" % "finatra" % "4.3.2",
+
+        "com.twitter.common.zookeeper" % "server-set" % "1.0.7",
+
+        "com.twitter" % "finagle-serversets" % "5.0.0",
```
Comment (johanoskarsson) [inline project/Project.scala]: Have you tried pulling in the same version of finagle that we use for the rest of the project? Too different?

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/adapter/JsonAdapter.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+object JsonAdapter extends Adapter {
+  type annotationType = Annotation
+  type annotationTypeType = AnnotationType
+  type binaryAnnotationType = JsonBinaryAnnotation
+  type endpointType = Endpoint
+  type spanType = JsonSpan
+  type traceSummaryType = JsonTraceSummary
+
+  def apply(a: annotationType): Annotation = a
+  //def apply(a: Annotation): annotationType = a
```
Comment (johanoskarsson) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/adapter/JsonAdapter.scala]: Can these be removed?

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+  get("/static") { request =>
+    render(path = "static.mustache", exports = new StaticObject)
+  }
+
+  get("/api/query") { request =>
+    /* Get trace ids */
+    val traceIds = QueryRequest(request) match {
+      case r: SpanQueryRequest => {
+        log.debug(r.toString)
```
Comment (johanoskarsson) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala]: Move out the logging to before the match?

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+import com.twitter.util.Future
+import com.twitter.zipkin.adapter.{JsonQueryAdapter, JsonAdapter, ThriftQueryAdapter, ThriftAdapter}
+import com.twitter.zipkin.gen
+import com.twitter.zipkin.config.ZipkinWebConfig
+import java.nio.ByteBuffer
+import java.text.SimpleDateFormat
+import java.util.Calendar
+import org.jboss.netty.handler.codec.http.HttpResponse
+
+class App(config: ZipkinWebConfig, client: gen.ZipkinQuery.FinagledClient) extends FinatraApp {
```
Comment (johanoskarsson) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala:35]: I'd suggest adding some comments to this class. Maybe even describe the use and purpose of each endpoint?

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/ZipkinWeb.scala [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  var server: Option[Server] = None
+
+  def start() {
+    val clientBuilder = ClientBuilder()
+      .codec(ThriftClientFramedCodec())
+      .hostConnectionLimit(1)
+
+    val clientService = config.zkClient match {
+      case None => {
+        clientBuilder.hosts(new InetSocketAddress("localhost", 3002))
```
Comment (johanoskarsson) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/ZipkinWeb.scala]: Shouldn't this be in a config and not hardcoded?

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/QueryRequest.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
+package com.twitter.zipkin.web
+
+import com.twitter.zipkin.gen
+import java.text.SimpleDateFormat
+import com.twitter.util.Time
+import com.posterous.finatra.Request
+
+object QueryRequest {
```
Comment (johanoskarsson) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/QueryRequest.scala:8]: Would be nice with a few comments in here. Maybe describing the type of requests or just what happens in the apply method

Comment (franklinhu) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/QueryRequest.scala:8]: Fixed in 1dd303552a6664a5867eba23ef13369bd9a0a266

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/ZipkinWeb.scala (hunk 5 above)
Comment (franklinhu) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/ZipkinWeb.scala]: Fixed in    56a6f08

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala (hunk 4 above)
Comment (franklinhu) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala:35]: Fixed in    729df15

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala (hunk 3 above)
Comment (franklinhu) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/web/App.scala]: Fixed in 729df15

File: zipkin-finatra/src/main/scala/com/twitter/zipkin/adapter/JsonAdapter.scala (hunk 2 above)
Comment (franklinhu) [inline zipkin-finatra/src/main/scala/com/twitter/zipkin/adapter/JsonAdapter.scala]: Fixed in    48e2efe

File: project/Project.scala (hunk 1 above)
Comment (franklinhu) [inline project/Project.scala]: Fixed in    6535894

**Comment by johanoskarsson** [comment]
+1

**Pull request closed without merging**